### PR TITLE
extended RFC 2617 to RFC 7616  support by adding SHA-256 and SHA-512-256 digest authentication.

### PR DESCRIPTION
**Pull request by xiaotianshi2**

### CODE REVIEW

File: libavformat/http.c [hunk 1]
```diff
@@ -1288,6 +1288,7 @@ static int http_connect(URLContext *h, const char *path, const char *local_path,
              proxyauthstr ? "Proxy-" : "", proxyauthstr ? proxyauthstr : "");
 
     av_log(h, AV_LOG_DEBUG, "request: %s\n", s->buffer);
+    //av_log(h, AV_LOG_INFO, "request: %s\n", s->buffer);
```
Comment (joepadmiraal) [inline libavformat/http.c]: Please remove this comment

File: libavformat/httpauth.c [hunk 2]
```diff
@@ -242,11 +401,27 @@ static char *make_digest_auth(HTTPAuthState *state, const char *username,
...
+    }
+    /* Unsupported algorithm */
+    return NULL;
+}
+
 char *ff_http_auth_create_response(HTTPAuthState *state, const char *auth,
                                    const char *path, const char *method)
 {
     char *authstr = NULL;
-
```
Comment (joepadmiraal) [inline libavformat/httpauth.c]: I'd like the code to be as similar to the upstream repository so please remove the removal this newline.

File: libavformat/httpauth.c [hunk 3]
```diff
@@ -276,7 +451,6 @@ char *ff_http_auth_create_response(HTTPAuthState *state, const char *auth,
         av_free(decoded_auth);
     } else if (state->auth_type == HTTP_AUTH_DIGEST) {
         char *username = ff_urldecode(auth), *password;
-
```
Comment (joepadmiraal) [inline libavformat/httpauth.c]: I'd like the code to be as similar to the upstream repository so please remove the removal this newline.

File: libavformat/httpauth.c [hunk 4]
```diff
@@ -199,6 +221,143 @@ static char *make_digest_auth(HTTPAuthState *state, const char *username,
...
+    if (!strcmp(digest->algorithm, "SHA-256") || !strcmp(digest->algorithm, "SHA-256-sess")) {
+        sha256ctx = av_sha_alloc();
+        if (!sha256ctx)
+            return NULL;
+        av_sha_init(sha256ctx, 256);
+        update_sha256_strings(sha256ctx, username, ":", state->realm, ":", password, NULL);
+        av_sha_final(sha256ctx, hash);
+        ff_data_to_hex(A1hash, hash, 32, 1);
+        A1hash[64] = 0;
+        if (!strcmp(digest->algorithm, "SHA-256-sess")) {
```
Comment (joepadmiraal) [inline libavformat/httpauth.c:302]: This seems to be overwriting A1hash. Should this not be an else statement so we write A1hash only once?

File: libavformat/dashenc_http.c [hunk 5]
```diff
@@ -313,7 +313,11 @@ static void *thr_io_close(connection *conn) {
     if (ret < 0 || response_code >= 500) {
         av_log(NULL, AV_LOG_INFO, "-event- request failed ret=%d, conn_nr: %d, response_code: %d, url: %s.\n", ret, conn->nr, response_code, conn->url);
         abort_if_needed(conn->must_succeed);
-        ff_format_io_close(conn->s, &conn->out);
+        //must check if this is NULL or it causes crash on segmentation fault due to NULL pointer
+        //check why conn->s (AVFormatContext) becomes NULL
```
Comment (joepadmiraal) [inline libavformat/dashenc_http.c:317]: Can you check if this does not happen when you use the http_persistent=1 option?
If so we can remove this code and I will create a ticket for supporting non persistent connections.

**Review by joepadmiraal — CHANGES_REQUESTED** [review]
Added some minor comments.

File: libavformat/httpauth.c (hunk 2 above)
Comment (xiaotianshi2) [reply libavformat/httpauth.c]: The line is added back.

File: libavformat/httpauth.c (hunk 3 above)
Comment (xiaotianshi2) [reply libavformat/httpauth.c]: The line is added back.

File: libavformat/httpauth.c (hunk 4 above)
Comment (xiaotianshi2) [reply libavformat/httpauth.c:302]: Not really. 
Initially:
A1Hash = username) ":" + realm) ":" + passwd  //must have for both SHA-256 and SHA-256-sess
If algorithm is "SHA-256-sess", 
A1hash = A1hash + ":" + nonce-prime) + ":" + cnonce //This is only needed for "SHA-256-sess"
So the implementation looks correct.

File: libavformat/dashenc_http.c (hunk 5 above)
Comment (xiaotianshi2) [reply libavformat/dashenc_http.c:317]: I'll check.

Comment (xiaotianshi2) [reply libavformat/dashenc_http.c:317]: I tested with http_persistent=1, it never gets segmentation fault or crashes.

**Review by joepadmiraal — APPROVED** [review]

**Pull request merged**